numpy/core/src/multiarray/descriptor.c

```diff
@@ -772,6 +772,60 @@ _is_tuple_of_integers(PyObject *obj)
     return 1;
 }
 
+static int
+validate_structured_object_dtype(PyArray_Descr *new, PyArray_Descr *conv)
+{
+    PyObject *names, *fields, *name, *tup;
+    PyArray_Descr *dtype;
+    Py_ssize_t names_size, pos;
+
+    if (PyDataType_HASFIELDS(conv)) {
+        names = conv->names;
+        fields = conv->fields;
+        if (new->kind == 'O') {
+            /* 'O' can only be mapped to a single 'O' field */
+            names_size = PyTuple_GET_SIZE(names);
+            if (names_size != 1) {
+                goto fail;
+            }
+            name = PyTuple_GET_ITEM(names, 0);
+            if (name == NULL) {
+                return -1;
+            }
+            tup = PyDict_GetItem(fields, name);
+            if (tup == NULL) {
+                return -1;
+            }
+            dtype = (PyArray_Descr *)PyTuple_GET_ITEM(tup, 0);
+            if (dtype == NULL) {
+                return -1;
+            }
+            if (dtype->kind != 'O') {
+                goto fail;
+            }
+        }
+        else {
```

Comment:
Can this whole else block be replaced by 
```
if (PyDataType_REFCHK(conv)) {
    goto fail;
}
```
?

That would also catch the case there are objects in nested structures.